search-service/src/main/kotlin/com/egm/stellio/search/authorization/permission/service/PermissionService.kt

```diff
@@ -0,0 +1,307 @@
+        }.plus("modifiedAt" to ngsiLdDateTime())
+            .forEach {
+                when {
+                    it.key == "target" -> {
+                        val target = input["target"] as Map<String, Any>
+                        target["id"]?.let {
+                            updatePermissionAttribute(permissionId, "target_id", it).bind()
+                        }
+                    }
+
+                    listOf(
+                        "id",
+                        "assignee",
+                        "action",
+                        "assigner",
+                        "modifiedAt"
+                    ).contains(it.key) -> {
+                        val columnName = it.key
+                        val value = it.value
+                        updatePermissionAttribute(permissionId, columnName, value).bind()
+                    }
+
+                    else -> {
+                        BadRequestDataException("invalid permission parameter  ${it.key}")
+                            .left().bind<Unit>()
+                    }
+                }
+            }
+    }
+
+    private suspend fun updatePermissionAttribute(
+        permissionId: URI,
+        columnName: String,
+        value: Any?
+    ): Either<APIException, Unit> {
+        val updateStatement = Update.update(columnName, value)
+        return r2dbcEntityTemplate.update(Permission::class.java)
+            .matching(query(where("id").`is`(permissionId)))
+            .apply(updateStatement)
+            .map { Unit.right() }
+            .awaitFirst()
+    }
+
+    suspend fun getPermissions(
+        filters: PermissionFilters = PermissionFilters(),
+        limit: Int = Int.MAX_VALUE,
+        offset: Int = 0,
+    ): Either<APIException, List<Permission>> = either {
+        val filterQuery = buildWhereStatement(filters)
+
+        val selectStatement =
+            """
+            SELECT id,
+                target_id,
+                assignee,
+                action,
+                created_at,
+                modified_at,
+                assigner
+            FROM permission
+            WHERE $filterQuery
+            ORDER BY permission.created_at
+            LIMIT :limit
+            OFFSET :offset
+            """.trimIndent()
+
+        databaseClient.sql(selectStatement)
+            .bind("limit", limit)
+            .bind("offset", offset)
+            .allToMappedList { rowToPermission(it).bind() }
+    }
+
+    suspend fun getPermissionsCount(
+        filters: PermissionFilters = PermissionFilters(),
+    ): Either<APIException, Int> {
+        val filterQuery = buildWhereStatement(filters)
+
+        val selectStatement =
+            """
+            SELECT count(distinct permission.id)
+            FROM permission
+            WHERE $filterQuery
+            """.trimIndent()
+        return databaseClient.sql(selectStatement)
+            .oneToResult { toInt(it["count"]) }
+    }
+
+    companion object {
```

Comment:
what's the advantage of putting these private functions into a companion object?